crates/router/src/core/user_role.rs

```diff
@@ -115,173 +113,235 @@ pub async fn update_user_role(
-        &user_from_token.org_id,
-    )
-    .await
-    .change_context(UserErrors::InternalServerError)?;
+    // let user_role_to_be_updated = user_to_be_updated
+    //     .get_role_from_db_by_merchant_id(&state, &user_from_token.merchant_id)
+    //     .await
+    //     .to_not_found_response(UserErrors::InvalidRoleOperation)?;
 
-    if !role_to_be_updated.is_updatable() {
-        return Err(report!(UserErrors::InvalidRoleOperation)).attach_printable(format!(
-            "User role cannot be updated from {}",
-            role_to_be_updated.get_role_id()
-        ));
-    }
+    let mut is_updated = false;
 
-    state
+    let v2_user_role_to_be_updated = match state
         .store
-        .update_user_role_by_user_id_merchant_id(
+        .find_user_role_by_user_id_and_lineage(
             user_to_be_updated.get_user_id(),
-            &user_role_to_be_updated
-                .merchant_id
-                .ok_or(UserErrors::InternalServerError)
-                .attach_printable("merchant_id not found in user_role")?,
-            UserRoleUpdate::UpdateRole {
-                role_id: req.role_id.clone(),
-                modified_by: user_from_token.user_id,
-            },
-            UserRoleVersion::V1,
+            &user_from_token.org_id,
+            &user_from_token.merchant_id,
+            None,
+            UserRoleVersion::V2,
         )
         .await
-        .to_not_found_response(UserErrors::InvalidRoleOperation)
-        .attach_printable("User with given email is not found in the organization")?;
-
-    auth::blacklist::insert_user_in_blacklist(&state, user_to_be_updated.get_user_id()).await?;
+    {
+        Ok(user_role) => Some(user_role),
+        Err(e) => {
+            if e.current_context().is_db_not_found() {
+                None
+            } else {
+                return Err(UserErrors::InternalServerError.into());
+            }
+        }
+    };
 
-    Ok(ApplicationResponse::StatusOk)
-}
+    if let Some(user_role) = v2_user_role_to_be_updated {
+        let role_to_be_updated = roles::RoleInfo::from_role_id(
+            &state,
+            &user_role.role_id,
+            &user_from_token.merchant_id,
+            &user_from_token.org_id,
+        )
+        .await
+        .change_context(UserErrors::InternalServerError)?;
 
-pub async fn transfer_org_ownership(
-    state: SessionState,
-    user_from_token: auth::UserFromToken,
-    req: user_role_api::TransferOrgOwnershipRequest,
-    _req_state: ReqState,
-) -> UserResponse<user_api::DashboardEntryResponse> {
-    if user_from_token.role_id != consts::user_role::ROLE_ID_ORGANIZATION_ADMIN {
-        return Err(report!(UserErrors::InvalidRoleOperation)).attach_printable(format!(
-            "role_id = {} is not org_admin",
-            user_from_token.role_id
-        ));
-    }
+        if !role_to_be_updated.is_updatable() {
+            return Err(report!(UserErrors::InvalidRoleOperation)).attach_printable(format!(
```

Comment:
We should also check (entity type relation) hierarchy here? Whether the user can perform operation at his current role in the given hierarchy or not?

Reply:
Yes we should.